### DISCUSSION TITLE
Cowper-Symonds / Johnson-Cook model implementation

### DISCUSSION THREAD

**Discussion by thibaultbarret** [thelfer/tfel#587] [Q&A]
Bonjour,
Dans le cadre de mes travaux de thèse, j'aurai besoin d'implémenter des modèles viscoplastiques phénoménologiques de type multiplicatif qu'on peut écrire sous la forme : sigma = f(p) * f(dp/dt). Un terme qui dépend de la déformation plastique équivalente et un second qui dépend de la vitesse de déformation équivalente.
J'ai cherché sur le site mais je n'ai pas trouvé comment accéder à la vitesse de déformation plastique équivalente.
Merci d'avance.

**Comment by thelfer** [comment]
Bonjour Thibault,

Nous avons un DSL dédié aux lois **isotropes** de la formet `dp/dt = F(sigma, p)`.

Si votre loi est isotrope, partant de votre expression, `sigma = f(p) * fr(dp/dt)`, pouvez-vous re-exprimer votre loi sous la forme: `dp/dt = F(sigma, p) = fr^{-1}(sigma/f(p))` ?

**Reply by thelfer** [reply]
Sinon, nous avons bien une implémentation de Johnson-Cook disponible, mais elle est loin d'être optimale. Si la solution précédente ne convient pas, je vous proposerai quelque chose de plus satisfaisant.

**Reply by thibaultbarret** [reply]
Merci pour la réponse rapide.
Oui je peux réécrire la loi sous cette forme. 
En partant de l'exemple de la loi de Cowper-Symonds : 
dp/dt = D(R - 1)^n
avec R l'écrouissage isotrope (loi mixte Swift-Voce dans mon cas)

**Reply by thelfer** [reply]
Voici un exemple de loi de Lemaître:  [StrainHardeningCreep.zip](https://github.com/user-attachments/files/15969789/StrainHardeningCreep.zip)
Dits moi si ca vous convient.

**Reply by thibaultbarret** [reply]
Est-ce qu'à partir de la @Brick StandardElastoViscoPlasticity en définissant successivement R, dR_dp, vp, dvp_dp il serait possible d'implémenter ma loi ?

**Reply by thelfer** [reply]
C'est également possible ! Mais a priori, les performances seront meilleures avec un DSL spécialisé.

**Comment by thibaultbarret** [comment]
Bonjour,
J'ai cherché des informations sur le DSL IsotropicStrainHardeningMisesCreep il n'y a pas d'indications sur ce que doit contenir le block @FlowRule.
Que faut-il renseigner pour pouvoir implémenter la loi de Cowper-Symonds ?

**Reply by thelfer** [reply]
La documentation (en français) est [ici](https://thelfer.github.io/tfel/web/documents/mfront/behaviours.pdf) (à partir de la page 24).
Je vous ai envoyé un exemple d'implémentation. Il "suffit" de donner `f`, `df_dseq`, `df_dp`. Je n'ai pas la loi de Cowper-Symonds en tête, n'hésitez pas à me donner son expression, j'y jetterai un oeil.

**Comment by thibaultbarret** [comment]
J'ai réussi à implémenter la loi de Swift sans problème en définissant dans le block @FlowRule : 
$$f = \bar{\sigma} - \sigma_y(p)$$
avec : $$\sigma_y =K \left( \varepsilon_0 + p\right)^{n}$$
Les dérivées de $f$ par rapport à $\bar{\sigma}$ et $p$ apparaissent naturellement.
Par contre, pour rajouter la dépendance à la vitesse de déformation j'ai du louper quelque chose.
Le critère de plasticité devient 
$$f = \bar{\sigma} - \left[ K \left( \varepsilon_0 + p\right)^{n}\right]  \times  \left[ 1 + \left( \dfrac{\dot{p}}{D} \right)^{\frac{1}{e}}\right]$$.
Comment je dois exprimer $\dot{p}$ ?

**Reply by thelfer** [reply]
Il vaut écrire f =0 et en déduire dp_dt, puis intégrer cette expression dans le DSL IsotropicStrainHardeningMisesCreep

**Reply by thelfer** [reply]
Cependant, en écrivant cela pour p nul, on peut avoir un soucis à contrainte nulle... Hors c'est exactement ce qu'on aura à la première itération du premier pas de temps...

**Reply by thibaultbarret** [reply]
Si je reprends depuis le début : 
La loi que je veux implémenter est : 
$$f=\bar{\sigma} - \left[ K \left( \varepsilon_0 + p\right)^{n}\right]  \times  \left[ 1 + \left( \dfrac{\dot{p}}{D} \right)^{\frac{1}{e}}\right]$$
1. J'exprime $\dot{p}$ en fonction de $\bar{\sigma}$ et $p$ pour avoir $\dot{p} = f(\bar{\sigma}, p)$ : 
$$\dot{p}=\left[ D\left(\dfrac{\bar{\sigma}}{K \left( \varepsilon_{0} + p \right)^{n}}  - 1 \right) \right]^{e}$$
2. Je calcule les dérivées de $\dfrac{\partial f}{\partial\bar{\sigma}}$ et $\dfrac{\partial f}{\partial p}$
$$\dfrac{\partial f}{\partial\bar{\sigma}} =\dfrac{eD^{e} \left[ \bar{\sigma} - K \left( \varepsilon_{0} + p \right)^{n} \right]^{e-1}}{K^{e} \left( \varepsilon_{0} + p \right)^{en} }$$
$$\dfrac{\partial f}{\partial p} =\dfrac{e \bar{\sigma} n D^{e} \left[ \bar{\sigma} - K \left( \varepsilon_{0} + p \right)^{n} \right]^{e-1} \times \left(\varepsilon_{0} +p \right)^{-en-1}}{K^{e}}$$

La loi se compile, mais lorsque je la teste sur un essai de traction, elle échoue instantanément.
Dans le fichier que vous m'avez joins, il y a un $p0$ égal à 1e-6 qui sert à éviter les problèmes à la première itération ?

**Reply by thelfer** [reply]
Non, ici epsilon0 joue ce rôle. Le problème est que l'expression 1 peut conduire à des valeurs négatives...

Il faut peut-être prendre la valeur positive du terme de droite.

Par ailleurs, j'espère que l'exposant e est plus grand que 1

**Comment by thibaultbarret** [comment]
En navigant dans la Front Gallery, je suis tombé sur un exemple d'implémentation de la loi de Johnson-Cook : https://github.com/thelfer/MFrontGallery/blob/master/generic-behaviours/viscoplasticity/JohnsonCook_ssr.mfront
J'ai voulu m'en inspirer de l'implémentation mais dès que je rajoute la sensibilité à la vitesse de déformation, impossible de faire converger le test sur un point matériel. 
Sur l'exemple cité ci-dessus, je comprends la condition sur la valeur de epsilon0_dt pour éviter les problèmes avec la fonction log, mais dans mon cas c'est une puissance, donc pas de problème avec les valeurs négatives. 
[Swift-CowperSymonds.zip](https://github.com/user-attachments/files/16444975/Swift-CowperSymonds.zip)

**Reply by thelfer** [reply]
Bonjour @thibaultbarret,

Comme l'exposant `e` est plus petit que 1, il y a des problèmes numériques pour `dp=0`.  J'ai fait deux modifications dans le fichier ci-joint:

1. j'ai rajouté un petit epsilon pour éviter que `dCowperSymonds_ddp` n'explose.
2. j'ai corrigé le calcul de `dCowperSymonds_ddp`  (Il manquait une division par `coefD_dt`.

Avec ça, le cas test `traction01.mtest` tourne sans sous-découpage en 500 pas de temps. Je n'ai pas regardé les résultats.

La loi serait beaucoup plus efficace avec:

- une meilleure prédiction initiale de `dp`,
- une réduction à une équation scalaire.

On pourra regarder ça quand vous aurez valider les résultats.

Amicalement,

[Swift-CowperSymonds.zip](https://github.com/user-attachments/files/16452002/Swift-CowperSymonds.zip)

**Reply by thibaultbarret** [reply]
Bonjour @thelfer,

Effectivement, un manque de rigueur sur la dérivation du terme visqueux. 
Dès lors que j'enlève le petit epsilon dans le terme visqueux, ça ne converge pas.
J'ai validé l'implémentation par rapport à la solution analytique, cependant je voudrais vérifier la loi dans Abaqus également.

Une fois la loi compilé avec les options `--interface=generic` et `--obuild`, j'obtiens un librairie dynamique `libBehaviour.so`.
Je l'ai renommée pour qu'elle coincide avec les indications données : https://thelfer.github.io/tfel/web/abaqus.html
Mais j'ai le message d'erreur : undefined symbol: SWIFT_TRIDIMENSIONAL
Qu'est ce que je manque ?

**Reply by thelfer** [reply]
Pour utiliser la loi avec `Abaqus/Standard`, il faut utiliser l'interface `abaqus`: `--interface=abaqus`
Cela créera une librairie dynamique différente compatible avec ce solveur.

**Comment by thibaultbarret** [comment]
Bonjour @thelfer,

J'avais suivi les instructions pour une installation rapide de Mfront. Une fois réinstallé et en utilisant les bonnes options dans le cmake ça fonctionne parfaitement. L'implémentation est validée, j'obtiens les mêmes résultats qu'avec une loi tabulée et qu'avec Mtest.

Merci pour votre aide.

**Reply by thelfer** [reply]
Super ! Il y a certainement encore des choses à améliorer, mais c'est un bon début !

**Comment by thibaultbarret** [comment]
Bonjour @thelfer, je reviens vers vous.

La loi compilée fonctionne avec `Mtest` et `Abaqus/Standard` avec **1** élément mais dès lors que je bascule sur une simulation avec plusieurs milliers d'élément impossible de rentrer dans le régime plastique.

J'ai essayé la loi https://github.com/thelfer/MFrontGallery/blob/master/generic-behaviours/viscoplasticity/JohnsonCook_ssr.mfront dont j'étais partie et ça fonctionne. 
J'ai essayé en rajoutant le mot clé `@StrainMeasure Hencky` puisque je suis en `nlgeom` mais le calcul ne converge pas.

Le problème vient de l'opérateur tangent mais si j'ai bien compris, c'est interface qui s'en charge ?

**Reply by thelfer** [reply]
> J'ai essayé en rajoutant le mot clé @StrainMeasure Hencky puisque je suis en nlgeom mais le calcul ne converge pas.

Pour l'instant, ce n'est pas nécessaire. Si on ne le rajoute pas, on utilisera la formulation native d'Abaqus basé sur la dérivée objective de Jaumann.

> mais dès lors que je bascule sur une simulation avec plusieurs milliers d'élément impossible de rentrer dans le régime plastique.

C'est étrange. Les post-traitements vous montrent-ils que la contrainte seuil est dépassée ?

**Reply by thibaultbarret** [reply]
Avec des éléments à intégration complète (C3D8) non, avec des éléments à intégration réduite (C3D8R), la contrainte seuil est à peine dépassée.

**Reply by thelfer** [reply]
Donc il est normal d'avoir une réponse élastique non ?

**Reply by thibaultbarret** [reply]
Je voulais dire que lorsque le calcul s'arrête, il n'y a pas ou très peu de plasticité. L'incrément de temps tend vers 1e-9 et Abaqus stoppe le calcul.

**Reply by thelfer** [reply]
Effectivement, c'est plus clair ainsi. Vous pouvez partager votre implémentation ?

**Reply by thibaultbarret** [reply]
La voici, elle a été testée avec `mtest` et `Abaqus` sur une élément C3D8. 
[SwiftVoce-CowperSymonds-abq.mfront.zip](https://github.com/user-attachments/files/17368583/SwiftVoce-CowperSymonds-abq.mfront.zip)

**Reply by thelfer** [reply]
Deux détails:

 >  real SwiftVoce = pAlpha * coefK * pow(epsi0 + p_, expN) + (1-pAlpha)*(Q1*(1-Q2*exp(-Q3*max(p_eps, p))));
 
devrait être: `real SwiftVoce = pAlpha * coefK * pow(epsi0 + p_, expN) + (1-pAlpha)*(Q1*(1-Q2*exp(-Q3*max(p_eps, p_))));`
  
>   real CowperSymonds = 1. + pow(dp / coefD_dt, 1/expE);

`real CowperSymonds = 1. + pow(max(dp, strain(0)) / coefD_dt, 1/expE);`

**Reply by thelfer** [reply]
Could you compile your behaviour with the option `--@GenerateMTestFileOnFailure=true` as follows:

~~~~
mfront --@GenerateMTestFileOnFailure=true --obuild --interface=abaqus <your_file>.mfront
~~~~

This will generate an `mtest` file for each integration failure. Would you then post a couple ones here ?

**Reply by thibaultbarret** [reply]
Je joins 3 fichiers `MTest`. Bizarrement ils sont générés uniquement lorsque je teste la loi de comportement avec `MTest`.  Lors de l'exécution avec `Abaqus`, il n'y a pas de fichiers générés lorsque l'UMAT tourne.
Les modifications apportées au fichier `mfront` n'ont pas réglé le problème. 
J'ai testé la loi de comportement sur un essai de traction avec un nombre croissant d'éléments jusqu'à 1000, avec des `C3D8` et `C3D8R` et aucun problème, mais dès lors que je bascule sur une géométrie plus complexe comme une éprouvette entaillée, ça ne converge pas.
[mtestFiles.zip](https://github.com/user-attachments/files/17378681/mtestFiles.zip)

Est-ce que le fait d'avoir compilé la loi avec Mfront-5 peut avoir une influence ?

**Reply by thelfer** [reply]
> Bizarrement ils sont générés uniquement lorsque je teste la loi de comportement avec MTest. Lors de l'exécution avec Abaqus, il n'y a pas de fichiers générés lorsque l'UMAT tourne.

C'est étrange. Abaqus indique-t'il explicitement que c'est l'intégration de la loi qui pose problème ?

**Reply by thelfer** [reply]
Le premier mtest que j'ai ouvert contient ça:

~~~~
@Stress {981.58086152331,9.6541354114221e-08,9.654110891773e-08,0,0,0};
@Strain {0.4986,-0.2483651622024,-0.2483651622024,0,0,0};
~~~~

Est-ce normal d'aller à de tels niveau de contraintes ?

**Reply by thibaultbarret** [reply]
Abaqus indique qu'il y a atteint le maximum de tentatives. L'incrément de temps a atteint la limite que j'ai fixé à savoir 1e-9.
Lorsque le calcul se stoppe, la contrainte équivalente de Von Mises correspond à la limite d'élasticité.

Oui, mon application est la mise en forme de tôles métalliques, c'est tout à fait normal.

**Reply by thelfer** [reply]
A chaque fois qu'`Abaqus` sous-découpe, il doit dire pourquoi. Il semble que ce ne soit pas lié à une erreur dans la loi de comportement puisqu'il n'y a pas de fichiers `mtest` générés.
Mais je ne vois pas ce que ça peut être, mais ça nous aiderait de le savoir.

> Oui, mon application est la mise en forme de tôles métalliques, c'est tout à fait normal.

Ok, donc je vais regarder ce fichier pour voir ce qu'il se passe

**Reply by thelfer** [reply]
Les fichiers `mtest` passent sur ma machine mais de justesse. En calculant comparant la valeur calculée de `df_ddp` à une approximation numérique montre une nette différence: 

~~~~
6.95183 1e-12
dfp_ddp :
-7.24265
ndfp_ddp :
-0.290823
dfp_ddp - ndfp_ddp :
-6.95183
~~~~

Pour avoir cet affichage, j'ai compilé le fichier ainsi:

~~~~
$ mfront --obuild --interface=abaqus --@CompareToNumericalJacobian=true --@PerturbationValueForNumericalJacobianComputation=1e-8 --debug SwiftVoce-CowperSymonds-abq.mfront 
~~~~

**Reply by thibaultbarret** [reply]
J'ai trouvé le problème ...
Dans le terme `dSwfitVoce_ddp = theta * expN * coefK * pow(epsi0 + p_, expN - 1) + theta * (1-pAlpha)*(Q1*Q2*Q3*exp(Q3*p_));`
Je n'ai pas mis le signe `-` devant le Q3.
En apportant cette modification ça fonctionne ... parfaitement.
Je suis surpris que malgré cette erreur dans la dérivée du terme d'écrouissage, pour des cas simple la loi compilée fonctionne.
Je m'excuse pour cette erreur très bête 😅.

**Reply by thelfer** [reply]
Si vous avez de jolies images/vidéos, n'hésitez pas !

**Reply by thelfer** [reply]
> Je suis surpris que malgré cette erreur dans la dérivée du terme d'écrouissage, pour des cas simple la loi compilée fonctionne.
Je m'excuse pour cette erreur très bête

Avec un peu d'expérience, on comprend qu'il est important de compiler la loi en mode `debug`: on voit tout de suite à la convergence du Newton si on a fait une bêtise :)

**Reply by thibaultbarret** [reply]
Est-ce que c'est normal que `dfp_ddp - ndfp_ddp` ne soit pas du même ordre de grandeur ?

**Reply by thelfer** [reply]
Quand on se trompe, oui :)

**Reply by thelfer** [reply]
Plus sérieusement, il faut surtout vérifier à convergence du Newton: les premières itérations peuvent être perturbées (surtout la première).

**Reply by thibaultbarret** [reply]
Avec `mtest` j'ai : 

```SwiftVoceCowperSymonds::integrate() : beginning of resolution
SwiftVoceCowperSymonds::integrate() : iteration 0 : 4.12835e-05
6.57807e-12 3.6e-12
dfeel_ddeel :
[[1,0,0,0,0,0]
 [0,1,0,0,0,0]
 [0,0,1,0,0,0]
 [0,0,0,1,0,0]
 [0,0,0,0,1,0]
 [0,0,0,0,0,1]]
ndfeel_ddeel :
[[1,0,0,0,0,0]
 [0,1,0,0,0,0]
 [0,0,1,0,0,0]
 [0,0,0,1,0,0]
 [0,0,0,0,1,0]
 [0,0,0,0,0,1]]
dfeel_ddeel - ndfeel_ddeel :
[[-2.19269e-12,0,0,0,0,0]
 [0,-2.19269e-12,0,0,0,0]
 [0,0,-2.19269e-12,0,0,0]
 [0,0,0,0,0,0]
 [0,0,0,0,0,0]
 [0,0,0,0,0,0]]

4.38555e-12 6e-13
dfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
ndfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
dfeel_ddp - ndfeel_ddp :
[ -2.19269e-12 1.09951e-12 1.09335e-12 0 0 0 ]

7.45635e-10 6e-13
dfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
ndfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
dfp_ddeel - ndfp_ddeel :
[ -3.45712e-10 1.99964e-10 1.99959e-10 0 0 0 ]

1322.74 1e-13
dfp_ddp :
-2223.09
ndfp_ddp :
-3545.83
dfp_ddp - ndfp_ddp :
1322.74

SwiftVoceCowperSymonds::integrate() : iteration 1 : 4.87822e-05
6.63132e-12 3.6e-12
dfeel_ddeel :
[[1,5.879e-21,-5.879e-21,0,0,0]
 [5.879e-21,1.00002,-1.98575e-05,0,0,0]
 [-5.879e-21,-1.98575e-05,1.00002,0,0,0]
 [0,0,0,1.00004,0,0]
 [0,0,0,0,1.00004,0]
 [0,0,0,0,0,1.00004]]
ndfeel_ddeel :
[[1,0,1.88237e-14,0,0,0]
 [9.41186e-15,1.00002,-1.98575e-05,0,0,0]
 [9.41186e-15,-1.98575e-05,1.00002,0,0,0]
 [0,0,0,1.00004,0,0]
 [0,0,0,0,1.00004,0]
 [0,0,0,0,0,1.00004]]
dfeel_ddeel - ndfeel_ddeel :
[[-2.19269e-12,5.879e-21,-1.88237e-14,0,0,0]
 [-9.41185e-15,-2.19891e-12,1.5835e-15,0,0,0]
 [-9.41186e-15,1.5835e-15,-2.19891e-12,0,0,0]
 [0,0,0,0,0,0]
 [0,0,0,0,0,0]
 [0,0,0,0,0,0]]

8.87814e-10 6e-13
dfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
ndfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
dfp_ddeel - ndfp_ddeel :
[ -3.45712e-10 -9.81945e-11 4.43907e-10 0 0 0 ]

0.000582187 1e-13
dfp_ddp :
-65.5419
ndfp_ddp :
-65.5413
dfp_ddp - ndfp_ddp :
-0.000582187

SwiftVoceCowperSymonds::integrate() : iteration 2 : 3.8112e-05
7.6263e-12 3.6e-12
dfeel_ddeel :
[[1,2.75442e-19,-1.83628e-19,0,0,0]
 [2.75442e-19,1.00062,-0.000620241,0,0,0]
 [-1.83628e-19,-0.000620241,1.00062,0,0,0]
 [0,0,0,1.00124,0,0]
 [0,0,0,0,1.00124,0]
 [0,0,0,0,0,1.00124]]
ndfeel_ddeel :
[[1,0,0,0,0,0]
 [-2.93623e-13,1.00062,-0.000620241,0,0,0]
 [-2.93623e-13,-0.000620241,1.00062,0,0,0]
 [0,0,0,1.00124,0,0]
 [0,0,0,0,1.00124,0]
 [0,0,0,0,0,1.00124]]
dfeel_ddeel - ndfeel_ddeel :
[[-2.19269e-12,2.75442e-19,-1.83628e-19,0,0,0]
 [2.93624e-13,-2.5282e-12,3.35447e-13,0,0,0]
 [2.93623e-13,4.1824e-14,-1.94089e-12,0,0,0]
 [0,0,0,0,0,0]
 [0,0,0,0,0,0]
 [0,0,0,0,0,0]]

6.96831e-13 6e-13
dfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
ndfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
dfeel_ddp - ndfeel_ddp :
[ 3.48499e-13 -1.74138e-13 -1.74194e-13 0 0 0 ]

3.92777e-10 6e-13
dfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
ndfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
dfp_ddeel - ndfp_ddeel :
[ 1.96389e-10 -9.81946e-11 -9.81942e-11 0 0 0 ]

0.00103743 1e-13
dfp_ddp :
-6.17951
ndfp_ddp :
-6.17848
dfp_ddp - ndfp_ddp :
-0.00103743

SwiftVoceCowperSymonds::integrate() : iteration 3 : 1.85593e-05
1.58302e-11 3.6e-12
dfeel_ddeel :
[[1,1.45903e-18,-1.45903e-18,0,0,0]
 [1.45903e-18,1.00493,-0.00492815,0,0,0]
 [-1.45903e-18,-0.00492815,1.00493,0,0,0]
 [0,0,0,1.00986,0,0]
 [0,0,0,0,1.00986,0]
 [0,0,0,0,0,1.00986]]
ndfeel_ddeel :
[[1,0,-4.62619e-12,0,0,0]
 [0,1.00493,-0.00492815,0,0,0]
 [0,-0.00492815,1.00493,0,0,0]
 [0,0,0,1.00986,0,0]
 [0,0,0,0,1.00986,0]
 [0,0,0,0,0,1.00986]]
dfeel_ddeel - ndfeel_ddeel :
[[2.43361e-12,1.45903e-18,4.62618e-12,0,0,0]
 [1.45903e-18,1.43663e-12,4.38489e-12,0,0,0]
 [-1.45903e-18,2.07179e-12,-8.7641e-13,0,0,0]
 [0,0,0,2.22045e-16,0,0]
 [0,0,0,0,2.22045e-16,0]
 [0,0,0,0,0,2.22045e-16]]

4.38538e-12 6e-13
dfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
ndfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
dfeel_ddp - ndfeel_ddp :
[ -2.19269e-12 1.09635e-12 1.09635e-12 0 0 0 ]

5.42101e-10 6e-13
dfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
ndfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
dfp_ddeel - ndfp_ddeel :
[ -3.45712e-10 -9.81945e-11 -9.81942e-11 0 0 0 ]

0.00105856 1e-13
dfp_ddp :
-1.50031
ndfp_ddp :
-1.49925
dfp_ddp - ndfp_ddp :
-0.00105856

SwiftVoceCowperSymonds::integrate() : iteration 4 : 2.63722e-06
4.51695e-11 3.6e-12
dfeel_ddeel :
[[1,3.21009e-18,-4.01261e-18,0,0,0]
 [3.21009e-18,1.01084,-0.0108427,0,0,0]
 [-4.01261e-18,-0.0108427,1.01084,0,0,0]
 [0,0,0,1.02169,0,0]
 [0,0,0,0,1.02169,0]
 [0,0,0,0,0,1.02169]]
ndfeel_ddeel :
[[1,-5.03024e-12,0,0,0,0]
 [-5.03024e-12,1.01084,-0.0108427,0,0,0]
 [-2.51512e-12,-0.0108427,1.01084,0,0,0]
 [0,0,0,1.02169,0,0]
 [0,0,0,0,1.02169,0]
 [0,0,0,0,0,1.02169]]
dfeel_ddeel - ndfeel_ddeel :
[[-2.19269e-12,5.03025e-12,-4.01261e-18,0,0,0]
 [5.03025e-12,-1.09908e-11,7.15583e-12,0,0,0]
 [2.51512e-12,-1.045e-11,-1.80322e-12,0,0,0]
 [0,0,0,4.44089e-16,0,0]
 [0,0,0,0,4.44089e-16,0]
 [0,0,0,0,0,4.44089e-16]]

4.38544e-12 6e-13
dfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
ndfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
dfeel_ddp - ndfeel_ddp :
[ -2.19269e-12 1.09635e-12 1.0964e-12 0 0 0 ]

8.11458e-10 6e-13
dfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
ndfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
dfp_ddeel - ndfp_ddeel :
[ -7.805e-11 -3.65857e-10 -3.67551e-10 0 0 0 ]

0.00107073 1e-13
dfp_ddp :
-0.881346
ndfp_ddp :
-0.880276
dfp_ddp - ndfp_ddp :
-0.00107073

SwiftVoceCowperSymonds::integrate() : iteration 5 : 3.84353e-08
3.05499e-11 3.6e-12
dfeel_ddeel :
[[1,5.30865e-18,-4.42387e-18,0,0,0]
 [5.30865e-18,1.01195,-0.011954,0,0,0]
 [-4.42387e-18,-0.011954,1.01195,0,0,0]
 [0,0,0,1.02391,0,0]
 [0,0,0,0,1.02391,0]
 [0,0,0,0,0,1.02391]]
ndfeel_ddeel :
[[1,0,-1.10675e-11,0,0,0]
 [0,1.01195,-0.011954,0,0,0]
 [5.53377e-12,-0.011954,1.01195,0,0,0]
 [0,0,0,1.02391,0,0]
 [0,0,0,0,1.02391,0]
 [0,0,0,0,0,1.02391]]
dfeel_ddeel - ndfeel_ddeel :
[[4.58367e-12,5.30865e-18,1.10675e-11,0,0,0]
 [5.30865e-18,-4.73888e-12,-2.98768e-12,0,0,0]
 [-5.53377e-12,-2.20796e-13,1.4162e-12,0,0,0]
 [0,0,0,4.44089e-16,0,0]
 [0,0,0,0,4.44089e-16,0]
 [0,0,0,0,0,4.44089e-16]]

4.38544e-12 6e-13
dfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
ndfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
dfeel_ddp - ndfeel_ddp :
[ -2.19269e-12 1.09635e-12 1.0964e-12 0 0 0 ]

3.86213e-10 6e-13
dfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
ndfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
dfp_ddeel - ndfp_ddeel :
[ 1.93106e-10 -9.65535e-11 -9.6553e-11 0 0 0 ]

0.00107243 1e-13
dfp_ddp :
-0.82566
ndfp_ddp :
-0.824588
dfp_ddp - ndfp_ddp :
-0.00107243

SwiftVoceCowperSymonds::integrate() : iteration 6 : 2.82935e-11
3.10098e-11 3.6e-12
dfeel_ddeel :
[[1,4.43005e-18,-4.43005e-18,0,0,0]
 [4.43005e-18,1.01197,-0.0119707,0,0,0]
 [-4.43005e-18,-0.0119707,1.01197,0,0,0]
 [0,0,0,1.02394,0,0]
 [0,0,0,0,1.02394,0]
 [0,0,0,0,0,1.02394]]
ndfeel_ddeel :
[[1,0,1.10826e-11,0,0,0]
 [0,1.01197,-0.0119707,0,0,0]
 [0,-0.0119707,1.01197,0,0,0]
 [0,0,0,1.02394,0,0]
 [0,0,0,0,1.02394,0]
 [0,0,0,0,0,1.02394]]
dfeel_ddeel - ndfeel_ddeel :
[[-2.19269e-12,4.43005e-18,-1.10826e-11,0,0,0]
 [4.43005e-18,1.93978e-12,7.56746e-12,0,0,0]
 [-4.43005e-18,-7.4451e-13,7.48135e-12,0,0,0]
 [0,0,0,4.44089e-16,0,0]
 [0,0,0,0,4.44089e-16,0]
 [0,0,0,0,0,4.44089e-16]]

4.3856e-12 6e-13
dfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
ndfeel_ddp :
[ 1 -0.5 -0.5 0 0 0 ]
dfeel_ddp - ndfeel_ddp :
[ -2.1928e-12 1.09646e-12 1.09635e-12 0 0 0 ]

4.63776e-10 6e-13
dfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
ndfp_ddeel :
[ 0.769231 -0.384615 -0.384615 0 0 0 ]
dfp_ddeel - ndfp_ddeel :
[ 1.93093e-10 -9.65466e-11 1.74137e-10 0 0 0 ]

0.00107246 1e-13
dfp_ddp :
-0.824891
ndfp_ddp :
-0.823818
dfp_ddp - ndfp_ddp :
-0.00107246

SwiftVoceCowperSymonds::integrate() : iteration 7 : 1.53389e-14
SwiftVoceCowperSymonds::integrate() : convergence after 7 iteration
```

Sur la première itération il y a une grande différence, à la fin, la différence est de l'ordre de 1e-4 ce qui est satisfaisant ?

**Reply by thelfer** [reply]
Oui. A la première itération, on a `dp=0`, donc la perturbation va conduire à évaluer le système pour une valeur négative de `dp`, et du coup, le `max(dp, 0)` pause soucis.
A première vue, tout est ok.

**Reply by thibaultbarret** [reply]
Si la loi fonctionne beaucoup mieux avec `MTest`, mais ça ne fonctionne pas avec une éprouvette entaillée. 
Abaqus renvoie ce message dans le fichier `.msg`

```
***NOTE: USER SUBROUTINE REQUESTS A TIME INCREMENT CHANGE.


 ***ERROR: TOO MANY ATTEMPTS MADE FOR THIS INCREMENT

 ***ERROR: THE ANALYSIS HAS BEEN TERMINATED DUE TO PREVIOUS ERRORS. ALL OUTPUT
           REQUESTS HAVE BEEN WRITTEN FOR THE LAST CONVERGED INCREMENT.
```

Dès lors que j'enlève la partie visqueuse de la loi de comportement tout se passe bien. Pourtant, Mfront indique qu'il a convergé dès la première itération lors de son utilisation.

**Reply by thelfer** [reply]
Il faut regarder ici: <https://github.com/thelfer/tfel/blob/master/mfront/include/MFront/Abaqus/AbaqusBehaviourHandler.hxx>
en particulier les lignes comprises entre 257 et 281 où les erreurs d'intégration sont traitées. Ce sont uniquement des erreurs d'intégration qui peuvent conduire à demander un sous-découpage du pas de temps dans votre cas.

Cependant, on voit que si on compile la loi avec l'option `--@GenerateMTestFileOnFailure=true`, on doit avoir la génération d'un fichier `mtest`. On peut regarder dans le source généré `src/abaqusSwiftVoceCowperSymonds.cxx`:

```c++
if(*PNEWDT<1){
try{
mfront::UmatSmallStrainMTestFileGenerator mg("abaqus","ABAQUSBEHAVIOUR","SWIFTVOCECOWPERSYMONDS");
mg.setModellingHypothesis(h);
const auto TVectorSize = mg.getTVectorSize();
const auto StensorSize = mg.getStensorSize();
const auto TensorSize  = mg.getTensorSize();
const auto dt = *DTIME>0 ? *DTIME : 1.e-50;
mg.setHandleThermalExpansion(false);
mg.addTime(0.);
```

Le fait qu'il n'y ait pas de `mtest` est incohérent

**Reply by thibaultbarret** [reply]
J'ai réessayer aujourd'hui, la loi fonctionne sur des cas de traction et de cisaillement mais dès lors que je passe sur une autre éprouvette impossible de rentrer dans le régime plastique. 
Est-ce qu'il y a un élément qui pourrait dégrader la matrice tangente renvoyé à Abaqus ?

**Reply by thelfer** [reply]
Le mieux serait de faire une visio rapide. Lundi matin ?

**Reply by thibaultbarret** [reply]
Je vous contacte par mail.
